xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,394 @@
+      nonce = lastResponse.getNonce();
+      versionInfo = lastResponse.getVersionInfo();
+    }
+    DiscoveryRequest.Builder builder =
+        DiscoveryRequest.newBuilder()
+            .setTypeUrl(SECRET_TYPE_URL)
+            .setResponseNonce(nonce)
+            .setVersionInfo(versionInfo)
+            .addResourceNames(sdsSecretConfig.getName())
+            .setErrorDetail(
+                com.google.rpc.Status.newBuilder()
+                    .setCode(errorCode)
+                    .setMessage(errorMessage)
+                    .build())
+            .setNode(clientNode);
+
+    DiscoveryRequest req = builder.build();
+    synchronized (requestObserver) {
+      requestObserver.onNext(req);
+    }
+  }
+
+  private void sendErrorToWatchers(final Throwable t) {
+    watcherExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            synchronized (watchers) {
+              for (SecretWatcher secretWatcher : watchers) {
+                try {
+                  secretWatcher.onError(Status.fromThrowable(t));
+                } catch (Throwable throwable) {
+                  logger.log(Level.SEVERE, "exception from onError", throwable);
+                }
+              }
+            }
+          }
+        });
+  }
+
+  private boolean processSecretsFromDiscoveryResponse(DiscoveryResponse response) {
+    List<Any> resources = response.getResourcesList();
+    checkState(resources.size() == 1, "exactly one resource expected");
+    boolean noException = true;
+    for (Any any : resources) {
+      final String typeUrl = any.getTypeUrl();
+      checkState(SECRET_TYPE_URL.equals(typeUrl), "wrong value for typeUrl %s", typeUrl);
+      Secret secret = null;
+      try {
+        secret = Secret.parseFrom(any.getValue());
+        if (!processSecret(secret)) {
+          noException = false;
+        }
+      } catch (InvalidProtocolBufferException e) {
+        logger.log(Level.SEVERE, "exception from parseFrom", e);
+      }
+    }
+    return noException;
+  }
+
+  private boolean processSecret(Secret secret) {
+    checkState(
+        sdsSecretConfig.getName().equals(secret.getName()),
+        "expected secret name %s",
+        sdsSecretConfig.getName());
+    boolean noException = true;
+    synchronized (watchers) {
+      for (SecretWatcher secretWatcher : watchers) {
+        try {
+          secretWatcher.onSecretChanged(secret);
+        } catch (Throwable throwable) {
+          noException = false;
+          logger.log(Level.SEVERE, "exception from onSecretChanged", throwable);
+        }
+      }
+    }
+    return noException;
+  }
+
+  /** Registers a secret watcher for this client's SdsSecretConfig. */
+  void watchSecret(SecretWatcher secretWatcher) throws InvalidProtocolBufferException {
+    checkNotNull(secretWatcher, "secretWatcher");
+    synchronized (watchers) {
+      checkState(watchers.add(secretWatcher), "duplicate secretWatcher addition");
```

Comment:
In this case the same `secretWatcher` value is passed the second time and I am just letting the caller know of their error. Since the value is the same (as the old value) it is still going to call the same watcher so I don't really see a problem. Otherwise what do you want the expected behavior to be?

Reply:
oops yes you are right. i don't know what i was thinking =p